

## 1. What is scHiC-SSM?

Single-cell Hi-C (scHi-C) profiles the chromatin three-dimensional (3D) conformation at single cell resolution, thus revealing genomic interactions with potential impact in regulating cell identities. However, high dimensionality and sparsity of scHi-C data often complicate the analysis. Here, we introduce a method, scHiC-SSM, for embedding scHi-C data that combines a semi-supervised deep generative model to learn latent features that accurately characterize scHi-C data. scHiC-SSM outperforms other tools in typical aspects of scHi-C data analysis, including cell embedding and cell type clustering on four independent scHi-C datasets. Taken together, scHiC-SSM represents a powerful tool to facilitate the study of 3D chromatin organization.

![image](https://github.com/LyuHaoUZH/scHiC-SSM/blob/main/img/flowchart.png)

## 2. Environment setup
We recommend you to create virtual environments by [anaconda](https://docs.anaconda.com/anaconda/install/linux/). Also, make sure you have an NVIDIA GPU with Linux x86_64 Driver Version >= 470.103.01 (compatible with CUDA Version 11.3) if you want to accelarate training process.
#### 2.1 Install by conda

```powershell
git clone https://github.com/LyuHaoUZH/scHiC-SSM
conda env create -f scHiC-SSM_conda_environment.yml
```
#### 2.2 Install by pip
Alternatively, you can install the package by pip.

```powershell
git clone https://github.com/LyuHaoUZH/scHiC-SSM
pip install -r python-requirements.txt
```
It is recommended to have pytorch installed after installing scHiC-SSM by using Conda:
```powershell
conda install pytorch torchvision torchaudio cudatoolkit=11.3 -c pytorch
```
Or pip:
```powershell
pip install torch torchvision torchaudio --extra-index-url https://download.pytorch.org/whl/cu113
```
## 3. Model training

```powershell
cd script

export INPUT_FILE=demo_data
export OUTPUT_FILE=test_result
export CELL_SUMMARY_FILE=supplementary_info/Li2019_Summary.txt
export GENOME_FILE=genome_info/hg19.chrom.sizes
export BAND=10 #Maximum genomic distance to be processed, a range of 10-100 is recommended.
export CHROMOSOME="chr1"
export RESOLUTION=1000000 #Resolution of scHi-C data.
export LATENT=10 #Dimension of latent space, a range of 10-100 is recommended.
export CPUNUM=10 #Number of CPUs to be used for parallel running.
export gpuFlag #(*selected) Use GPU or not. Default is False. Adding '-gpu' to turn it on.

python scHiC-SSM.py \
    --inPath INPUT_FOLDER \
    --outdir OUTPUT_FOLDER \
    --cellSummary CELL_SUMMARY_FILE \
    --genome GENOME_FILE \
    --bandMax BAND \
    --chromList CHROMOSOME \
    --resolution RESOLUTION \
    --nLatent LATENT \
    --parallelCPU CPUNUM \
    --verbose \
    -gpu (*selected)
```
## 4. Data access
Users can acquire [data](https://drive.google.com/drive/folders/1fcq1gKC1OO89tFd3bEEtAWgnveJesSyG?usp=sharing%20Dependencies) or extend scHiC-SSM to their own datasets.
## 5. Data format
#### 5.1 Input scHi-C data (tab separated file)
The input data is saved in the format of <chr1> <pos1> <chr2> <pos2> <counts>, without header
```powershell
chr1	3000000	chr1	3000000	39
chr1	4000000	chr1	4000000	33
chr1	5000000	chr1	5000000	36
chr1	5000000	chr1	6000000	3
chr1	6000000	chr1	6000000	54
chr1	4000000	chr1	7000000	1
chr1	7000000	chr1	7000000	36
chr1	7000000	chr1	8000000	2
chr1	8000000	chr1	8000000	33
chr1	4000000	chr1	9000000	1
chr1	6000000	chr1	9000000	1
chr1	7000000	chr1	9000000	1
chr1	8000000	chr1	9000000	1
chr1	9000000	chr1	9000000	43
chr1	7000000	chr1	10000000	1
```
#### 5.2 Output from scHiC-SSM
Under output directory:
1) pickle/scHiC-SSM_latentEmbeddings: full latent embeddings generated by scHiC-SSM
2) normalization/: normalized counts generated by scHiC-SSM
